doc/designs/quic-design/quic-api.md

```diff
@@ -818,6 +876,340 @@ reason string in bytes.
+ * any such call is a no-op.
+ *
+ * For QUIC:
+ *   Must be called only on a QUIC connection SSL object. Fails if called on a
+ *   stream object. Checks if a new stream has been created by the peer. If it
+ *   has, creates a new SSL object to represent it and returns it. Otherwise,
+ *   returns NULL. If multiple streams are available to be accepted, the oldest
+ *   stream (that is, the stream with the lowest stream ID) is accepted.
+ *
+ * For all other methods:
+ *   Returns NULL.
+ *
+ * The flags argument is unused and should be set to zero.
+ *
+ * To be MSMT-safe (i.e., can be called from multiple threads).
+ *
+ * If the QUIC connection SSL object is configured in blocking mode, this
+ * function will block unless the SSL_ACCEPT_STREAM_NO_BLOCK flag is passed.
+ *
+ * This function returns NULL if the effective incoming stream reject policy is
+ * `REJECT`.
+ */
+#define SSL_ACCEPT_STREAM_NO_BLOCK      1
+
+SSL *SSL_accept_stream(SSL *ssl, uint64_t flags);
+```
+
+#### `SSL_get_accept_stream_queue_len`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | No        | C             |
+
+```c
+/*
+ * Determine the number of streams waiting to be returned on a subsequent call
+ * to SSL_accept_stream. If this returns a non-zero value, the next call to
+ * SSL_accept_stream (on any thread) is guaranteed to work. Returns 0 for
+ * non-QUIC objects, or for QUIC stream objects.
+ *
+ * To be MSMT-safe.
+ */
+size_t SSL_get_accept_stream_queue_len(SSL *ssl);
+```
+
+#### `SSL_set_incoming_stream_reject_policy`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | No        | C             |
+
+```c
+/*
+ * Sets the policy for incoming streams. If `policy` is `AUTO` (the default):
+ *
+ *   - if `SSL_detach_stream` has been used, this is equivalent to `ACCEPT`;
+ *
+ *   - otherwise, if the default stream mode is
+ *     `SSL_DEFAULT_STREAM_MODE_AUTO_BIDI` or
+ *     `SSL_DEFAULT_STREAM_MODE_AUTO_UNI`, this is equivalent to `REJECT`;
+ *
+ *   - otherwise, this is equivalent to `ACCEPT`.
+ *
+ * If configured to `ACCEPT`, incoming streams are placed on the accept queue
+ * for application consumption. `aec` is ignored in this case.
+ *
+ * If configured to `REJECT`, incoming streams automatically have both their
+ * receiving and sending parts handled via non-normal termination. `aec` is an
+ * application error code used for the `STOP_SENDING` and `RESET_STREAM` frames
+ * used for the purposes of this termination. The default AEC value used if this
+ * function is never called is 0.
+ */
+#define SSL_INCOMING_STREAM_REJECT_POLICY_AUTO      0
+#define SSL_INCOMING_STREAM_REJECT_POLICY_ACCEPT    1
+#define SSL_INCOMING_STREAM_REJECT_POLICY_REJECT    2
+
+int SSL_set_incoming_stream_reject_policy(SSL *ssl, int policy, uint64_t aec);
```

Comment:
Why do we need this call? Or rather, should we have a call to set the values for MAX_STREAMS frame(s) and initial_max_streams transport parameters instead?

We could even set the default to support the single stream operation and require multi-stream apps to override it with something sensible for the concrete application.

Reply:
This is an interesting idea &mdash; use stream count flow control to prevent the peer from creating more streams.

In principle we could do this, so I suppose the question is if it's a better approach. 
My concern with this is that peers will probably assume `MAX_STREAMS` limits will clear eventually, rather than be indefinite. Let's take `quic-go` as an example: the `OpenStreamSync()` call creates a locally initiated stream if not limited by `MAX_STREAMS`. If limited by `MAX_STREAMS`, it waits indefinitely by default or with an optional timeout until a stream can be created. My guess is most users are going to assume it will return eventually, but if we do this, it wouldn't.

I think the issue here is, if we have connected to a peer and we are using single-stream mode, and that peer tries to open an additional stream, *something has gone wrong.* e.g. we have somehow gotten connected to an incompatible server. If we use `MAX_STREAMS` for this, it's quite plausible that the server will hang forever (or, hopefully, for some timeout period and then terminate the connection, but this may be optimistic). We don't want to defer discovery of this error condition, rather we want to signal it as soon as possible.

My basic view is, if we have gotten connected to the wrong thing and a peer is erroneously trying to create another stream, it's better if we signal that something has gone wrong *immediately* rather than have the peer waiting potentially forever for a increase in stream budget which will never come.

The other option, in addition to these two, would be to terminate the entire connection if a peer tries to open a remote stream, which would be an even more "fast and loud" way of failing if a peer tries to open another stream, which again indicates that something has gone quite wrong.

Reply:
OK, this makes sense. Do we need some separate API for handling MAX_STREAMS then?